### ISSUE TITLE
libffi-platypus-perl: FTBFS on s390x, ppc64, sparc64: test suite failures

### ISSUE THREAD

**Issue by gregoa** [plicease/FFI-Platypus#67]
We have the following bug reported to the Debian package of
FFI-Platypus (https://bugs.debian.org/829551):

It doesn't seem to be a bug in the packaging, so you may want to take
a look. Thanks!

The complete build logs are available at
https://buildd.debian.org/status/package.php?p=libffi-platypus-perl&suite=unstable

------8<-----------8<-----------8<-----------8<-----------8<-----

```
Package: libffi-platypus-perl
Version: 0.42-1
Severity: normal
Justification: never built previously

This package fails to build on 64-bit big endian architectures.
From the s390x build log:

  #   Failed test 'f0(22) = 22'
  #   at t/ffi_platypus_attach.t line 14.
  #          got: '0'
  #     expected: '22'
  
  #   Failed test 'f1(22) = 22'
  #   at t/ffi_platypus_attach.t line 15.
  #          got: '0'
  #     expected: '22'
  
  [...]
  
  Failed 25/57 test programs. 97/436 subtests failed.

-- 
Niko Tyni   ntyni@debian.org


```
------8<-----------8<-----------8<-----------8<-----------8<-----

Thanks for considering,
  gregor herrmann,
  Debian Perl Group

**Comment by plicease** [comment]
Thanks for the report.

Can you please provide the full build log, and the content of `include/ffi_platypus_probe.h`?

**Comment by gregoa** [comment]
The full build logs are available at https://buildd.debian.org/status/package.php?p=libffi-platypus-perl&suite=unstable for all architectures. Is this enough? Otherwise I could try to log into one of our porter machines, run the build and try to grab `include/ffi_platypus_probe.h` from there.

**Comment by gregoa** [comment]
Ok, that was easy. Here are the results from an s390x box:

```
$ cat include/ffi_platypus_probe.h                                                                                              
#ifndef FFI_PLATYPUS_PROBE_H
#define FFI_PLATYPUS_PROBE_H
#define FFI_PL_PROBE_ABI 1
#define FFI_PL_PROBE_COMPLEX 1
#define FFI_PL_PROBE_LONGDOUBLE 1
#endif
```

**Comment by plicease** [comment]
Thanks, that helps.  I am also including the probe data in the standard diagnostic so as to make this easier in the future.

**Comment by gregoa** [comment]
That's an excellent idea, thanks.

**Comment by gregoa** [comment]
I've uploaded a new package to Debian for unrelated reasons, and I've added a `cat include/ffi_platypus_probe.h` to our build script, so you can soon see the contents of this file on all our architectures in the build logs at the above mentioned URL.

**Comment by plicease** [comment]
Ah handy, thanks.

**Comment by plicease** [comment]
I've been able to reproduce this using QEMU, and fixed at least some of the errors.  When I have something workable I will release a dev version to CPAN which you can test.

**Comment by gregoa** [comment]
On Wed, 06 Jul 2016 07:08:32 -0700, Graham Ollis wrote:

> I've been able to reproduce this using QEMU, and fixed at least
> some of the errors. When I have something workable I will release a
> dev version to CPAN which you can test.

Excellent!
Just shout anytime if you'd like me to able a version to Debian.

Cheers,
gregor

-- 
 .''`.  Homepage https://info.comodo.priv.at/ - OpenPGP key 0xBB3A68018649AA06
 : :' : Debian GNU/Linux user, admin, and developer -  https://www.debian.org/
 `. `'  Member of VIBE!AT & SPI, fellow of the Free Software Foundation Europe
   `-

**Comment by plicease** [comment]
Try this:

**Comment by plicease** [comment]
oops.   Try this:

https://metacpan.org/pod/release/PLICEASE/FFI-Platypus-0.42_01/lib/FFI/Platypus.pm

I tested it under qemu s390x and ppc in addition to the usual native amd64 so my changes do not appear to have broken 32 bit bigendian.  I will be interested in if it works on real hardware.

**Comment by gregoa** [comment]
On Wed, 06 Jul 2016 12:11:44 -0700, Graham Ollis wrote:

> oops.   Try this:
> 
> https://metacpan.org/pod/release/PLICEASE/FFI-Platypus-0.42_01/lib/FFI/Platypus.pm
> 
> I tested it under qemu s390x and ppc in addition to the usual native amd64 so my changes do not appear to have broken 32 bit bigendian.  I will be interested in if it works on real hardware.

Great, thanks!
Uploaded to Debian, the build logs for 0.42.01-1 should soon appear at
https://buildd.debian.org/status/package.php?p=libffi-platypus-perl


Cheers,
gregor

-- 
 .''`.  Homepage https://info.comodo.priv.at/ - OpenPGP key 0xBB3A68018649AA06
 : :' : Debian GNU/Linux user, admin, and developer -  https://www.debian.org/
 `. `'  Member of VIBE!AT & SPI, fellow of the Free Software Foundation Europe
   `-

**Comment by plicease** [comment]
Matrix looks good.  Although am I crazy or is sh4 not running the test suite?

Unless I see any issues from CPAN testers I will likely do a production release in a couple of days.  I will close this bug then.

**Comment by gregoa** [comment]
On Wed, 06 Jul 2016 18:24:40 -0700, Graham Ollis wrote:

> Matrix looks good.  

Yay \o/

> Although am I crazy or is sh4 not running the test suite?

Interesting indeed. No idea what's going on there.
(But sh4 is a kind of experimental port.)
 
> Unless I see any issues from CPAN testers I will likely do a
> production release in a couple of days. I will close this bug then.

Sounds good, thank you.


Cheers,
gregor

-- 
 .''`.  Homepage https://info.comodo.priv.at/ - OpenPGP key 0xBB3A68018649AA06
 : :' : Debian GNU/Linux user, admin, and developer -  https://www.debian.org/
 `. `'  Member of VIBE!AT & SPI, fellow of the Free Software Foundation Europe
   `-

**Comment by plicease** [comment]
I believe this to be resolved in 0.43, now headed to CPAN.

**Comment by gregoa** [comment]
On Fri, 08 Jul 2016 00:33:18 -0700, Graham Ollis wrote:

> I believe this to be resolved in 0.43, now headed to CPAN.

Thanks, uploading to Debian right now.


Cheers,
gregor
-- 
 .''`.  Homepage https://info.comodo.priv.at/ - OpenPGP key 0xBB3A68018649AA06
 : :' : Debian GNU/Linux user, admin, and developer -  https://www.debian.org/
 `. `'  Member of VIBE!AT & SPI, fellow of the Free Software Foundation Europe
   `-
